### PR TITLE
[Resolver] Use improved bundler sorting algorithm

### LINKED ISSUE #1851: Gem::DependencyResolutionError when installing Rails 5.1.0.beta1
Hi,

There seems to be an issue when trying to install a pre release of Rails. This same issue also popped up for the first time (at least for me) when rails 5 beta was released, more info at #1653.

I get no output when I run `gem install rails --pre`. When I do run `gem install rails -v 5.1.0.beta1 --verbose`, I get this:

```
➜  gem install rails -v 5.1.0.beta1 --verbose
HEAD https://api.rubygems.org/api/v1/dependencies
200 OK
GET https://api.rubygems.org/api/v1/dependencies?gems=rails
200 OK
Getting SRV record failed: DNS result has no information for _rubygems._tcp.api.rubygems.org
GET https://api.rubygems.org/api/v1/dependencies?gems=actioncable,actionmailer,actionpack,actionview,activejob,activemodel,activerecord,activesupport,bundler,railties,sprockets-rails
200 OK
GET https://api.rubygems.org/api/v1/dependencies?gems=sprockets
200 OK
GET https://api.rubygems.org/api/v1/dependencies?gems=concurrent-ruby,rack
200 OK
GET https://api.rubygems.org/api/v1/dependencies?gems=i18n,minitest,nio4r,tzinfo,websocket-driver
200 OK
GET https://api.rubygems.org/api/v1/dependencies?gems=websocket-extensions
200 OK
GET https://api.rubygems.org/api/v1/dependencies?gems=builder,erubis,mail,rack-test,rails-dom-testing,rails-html-sanitizer
200 OK
GET https://api.rubygems.org/api/v1/dependencies?gems=method_source,nokogiri,rake,thor,thread_safe
200 OK
GET https://api.rubygems.org/api/v1/dependencies?gems=hike,journey,multi_json,rack-cache,rack-mount,tilt
200 OK
ERROR:  While executing gem ... (Gem::DependencyResolutionError)
    conflicting dependencies activesupport (= 3.0.0) and activesupport (= 5.1.0.beta1)
  Activated activesupport-5.1.0.beta1
  which does not match conflicting dependency (= 3.0.0)

  Conflicting dependency chains:
    rails (= 5.1.0.beta1), 5.1.0.beta1 activated, depends on
    activesupport (= 5.1.0.beta1), 5.1.0.beta1 activated

  versus:
    rails (= 5.1.0.beta1), 5.1.0.beta1 activated, depends on
    sprockets-rails (>= 2.0.0), 2.0.0 activated, depends on
    actionpack (>= 3.0), 3.0.0 activated, depends on
    activesupport (= 3.0.0)
```

I was, however able to install rails 5.1.0.beta1 using this snippet:
```
ruby -rbundler/inline -e "gemfile(true) do; source 'https://rubygems.org/'; gem 'rails', '5.1.0.beta1'; end"
```

Gem env

```
RubyGems Environment:
  - RUBYGEMS VERSION: 2.6.10
  - RUBY VERSION: 2.4.0 (2016-12-24 patchlevel 0) [x86_64-darwin15]
  - INSTALLATION DIRECTORY: /Users/millan/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0
  - USER INSTALLATION DIRECTORY: /Users/millan/.gem/ruby/2.4.0
  - RUBY EXECUTABLE: /Users/millan/.rbenv/versions/2.4.0/bin/ruby
  - EXECUTABLE DIRECTORY: /Users/millan/.rbenv/versions/2.4.0/bin
  - SPEC CACHE DIRECTORY: /Users/millan/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /Users/millan/.rbenv/versions/2.4.0/etc
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-darwin-15
  - GEM PATHS:
     - /Users/millan/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0
     - /Users/millan/.gem/ruby/2.4.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
     - "gem" => "--no-ri --no-rdoc"
  - REMOTE SOURCES:
     - https://rubygems.org/
```

### PR DESCRIPTION
**Pull request by segiddins**
# Description:

Fixes #1851 

# Tasks:

- [ ] Describe the problem / feature
- [ ] Write tests
- [ ] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: lib/rubygems/resolver.rb [hunk 1]
```diff
@@ -277,6 +277,40 @@ def allow_missing?(dependency)
...
+  # are given very negative values, so they _always_ sort first,
+  # before dependencies that are unconstrained
+  def amount_constrained(dependency)
+    @amount_constrained ||= {}
+    @amount_constrained[dependency.name] ||= begin
+      dependency_request_for_name = Gem::Resolver::DependencyRequest.new(Gem::Dependency.new(dependency.name), dependency.requester)
+      all = @set.find_all(dependency_request_for_name).size
+
+      if all <= 1
+        all - 1_000_000
```
Comment (drbrain) [inline lib/rubygems/resolver.rb]: where does this constant come from?

File: lib/rubygems/resolver.rb [hunk 2]
```diff
@@ -277,6 +277,40 @@ def allow_missing?(dependency)
...
+  # returns an integer \in (-\infty, 0]
+  # a number closer to 0 means the dependency is less constraining
+  #
+  # dependencies w/ 0 or 1 possibilities (ignoring version requirements)
+  # are given very negative values, so they _always_ sort first,
+  # before dependencies that are unconstrained
+  def amount_constrained(dependency)
+    @amount_constrained ||= {}
+    @amount_constrained[dependency.name] ||= begin
+      dependency_request_for_name = Gem::Resolver::DependencyRequest.new(Gem::Dependency.new(dependency.name), dependency.requester)
```
Comment (drbrain) [inline lib/rubygems/resolver.rb]: Can this become two lines? It's really long

File: lib/rubygems/resolver.rb (hunk 1 above)
Comment (djberg96) [inline lib/rubygems/resolver.rb]: @drbrain, it's not a constant, it's an obscure ruby-ism. 1_000 == 1000. They added it to make it easier to visualize large numbers iirc.

Comment (drbrain) [inline lib/rubygems/resolver.rb]: I mean, why is the value of the literal 1,000,000?

Comment (djberg96) [inline lib/rubygems/resolver.rb]: Oh, sorry, disregard. ;)

Comment (segiddins) [inline lib/rubygems/resolver.rb]: it's a sufficiently large "boost" for dependencies where the sum of all possibilities (even disregarding any requirements) is either 0 or 1, which I tried to explain in the comment above the method

Comment (drbrain) [inline lib/rubygems/resolver.rb]: How about extract it as `MIN_CONSTRAINT`? (`MAX_…`?)

**Comment by claudiob** [comment]
@segiddins Thanks for your change! I can confirm that this patch fixes the problem I reported in https://github.com/rubygems/rubygems/issues/1851#issuecomment-282444992

```sh
git:seg-resolver-sorting~/code/rubygems$ ruby -Ilib bin/gem install rails --pre -v 5.1.0.beta1
Fetching: activesupport-5.1.0.beta1.gem (100%)
Successfully installed activesupport-5.1.0.beta1
Fetching: actionview-5.1.0.beta1.gem (100%)
Successfully installed actionview-5.1.0.beta1
Fetching: actionpack-5.1.0.beta1.gem (100%)
Successfully installed actionpack-5.1.0.beta1
Fetching: activemodel-5.1.0.beta1.gem (100%)
Successfully installed activemodel-5.1.0.beta1
Fetching: activerecord-5.1.0.beta1.gem (100%)
Successfully installed activerecord-5.1.0.beta1
Fetching: activejob-5.1.0.beta1.gem (100%)
Successfully installed activejob-5.1.0.beta1
Fetching: actionmailer-5.1.0.beta1.gem (100%)
Successfully installed actionmailer-5.1.0.beta1
Fetching: actioncable-5.1.0.beta1.gem (100%)
Successfully installed actioncable-5.1.0.beta1
Fetching: railties-5.1.0.beta1.gem (100%)
Successfully installed railties-5.1.0.beta1
Fetching: rails-5.1.0.beta1.gem (100%)
Successfully installed rails-5.1.0.beta1
10 gems installed
```

**Comment by segiddins** [comment]
Done

**Comment by segiddins** [comment]
@homu r+

**Comment by dhh** [comment]
🙏. Would it be possible to push a rubygems-update with this fix? We're looking to release Rails 5.1.0 RC1 on Friday and it would be great if it would install cleanly with --pre. Thanks!

**Comment by segiddins** [comment]
I can try and backport this to 2.6.11 today, sure

**Comment by segiddins** [comment]
@dhh sorry for the delay, was busy studying for my last final exam. 2.6.11 was released today.

**Comment by dhh** [comment]
Awesome, thanks so much. Still hoping we can get Rails 5.1 RC1 out today. ❤️

**Comment by reza-naq** [comment]
@ddh @segiddins 
Hi there,
Weeks after the release of 5.1.0 and using rubygems 2.6.12 I have still the same problem.
Here are my Gemfile and the message I get when I run 'bundle update rails'
Gemfile:
_source 'https://rubygems.org'
ruby '2.4.1'
# gem 'rails', '~> 5.0.2'
gem 'rails', '~> 5.1.0'
gem 'puma', '~> 3.0'
gem 'byebug',            groups: [:development, :test, :staging], platforms: [:mri]

gem 'pg'
gem 'cancancan', '~> 1.10'
gem 'devise'      , github: 'plataformatec/devise'
gem 'simple_form' , github: 'plataformatec/simple_form'
gem 'devise-encryptable'
gem 'activemodel-serializers-xml'
gem 'globalize', git: 'https://github.com/globalize/globalize'
gem 'redcarpet'
gem 'country_select'

gem 'bootstrap-sass', '~> 3.3.6'
gem 'sass-rails', '~> 5.0'
gem 'coffee-rails'
gem 'uglifier', '>= 1.3.0'
gem 'jquery-rails'

gem 'turbolinks', '~> 5.0.0'
gem 'jbuilder', '~> 2.5'
gem 'redis', '~> 3.0'

gem 'haml-rails'
gem 'kaminari'
gem 'bcrypt'
gem 'friendly_id', '~> 5.1'
gem 'exception_notification'
gem 'geocoder'
gem 'pry-rails'
gem 'listen', '~> 3.1'

group :development, :test, :staging do
  gem 'letter_opener_web'
  gem 'i18n-tasks', '~> 0.9'
end

group :development do
  gem 'web-console'
  gem 'spring'
  gem 'spring-watcher-listen', '~> 2.0.0'
  gem "capistrano"
  gem "capistrano-git"
  gem "capistrano-bundler"
  gem "capistrano-rails"
  gem 'capistrano-rbenv'
end

group :test do
  gem 'sqlite3'
  gem "rspec-rails"
  gem 'cucumber-rails', :require => false
  gem 'aruba'
  gem 'database_cleaner'
  gem 'selenium-webdriver'
  gem 'factory_girl_rails'
  gem 'launchy', '~> 2.4.2'
  gem 'pickle', '>= 0.5'
  gem 'chronic'
  gem 'timecop'
  gem 'rails-controller-testing'
  gem 'capybara-screenshot'
end

gem 'sitemap_generator'
gem 'mini_magick'
gem 'carrierwave'
gem 'sidekiq'
gem "rack-mini-profiler"
gem 'record_tag_helper', '~> 1.0'
gem 'browser'
gem 'mini_portile2'

And this is the error message

Bundler could not find compatible versions for gem "actionpack":
  In Gemfile:
    rails (~> 5.1.0) java was resolved to 5.1.0, which depends on
      actionpack (= 5.1.0) java
    rails (~> 5.1.0) java was resolved to 5.1.0, which depends on
      actionpack (= 5.1.0)
    rails (~> 5.1.0) java was resolved to 5.1.0, which depends on
      actionpack (= 5.1.0) java
    rails (~> 5.1.0) java was resolved to 5.1.0, which depends on
      actionpack (= 5.1.0)
    simple_form java was resolved to 3.4.0, which depends on
      actionpack (< 5.1, > 4) java
    simple_form java was resolved to 3.4.0, which depends on
      actionpack (< 5.1, > 4)

Thanks

**Comment by dhh** [comment]
You hitting the dreaded upper-bound deadlock scenario. It's because
simple_form has a pessimistic upper-bound of less-than Rails 5.1. So it
won't resolve when you try to upgrade to Rails 5.1 because of that. You can
try to upgrade simple_form and see if they have a version out that lifts
the pessimistic restriction, but otherwise you'll have to fork the gem and
remove that restriction. Yes, that's totally ridiculous.

We are currently working on a broad community response to avoid such
needless deadlocking. Hope to have something ready soon. Sorry about this
mess!

**Comment by reza-naq** [comment]
@ddh
Many thanks for your prompt response! A proof of rails community vitality.

I'll try a change the upper-bound setting  in a local simple_form. I'll copy your message to the simple_form team. As you say the locking situation is pretty ridiculous. 

Do you think the other messages regarding 'actionpack' with no mention of simple_form is caused by the same pessimistic upper-bound approach?

**Comment by dhh** [comment]
Yes, what Bundler is saying is that Rails 5.1 wants actionpack 5.1, but
simple_form won't allow 5.1, so there's not a version of actionpack that
meets both those criteria, so it deadlocks and refuse to resolve.

I actually just removed the upper-bounds for a bunch of gems yesterday:
jbuilder, actionpack-xml_parser, etc. They all forced people to suffer the
same upgrade problems and for no good reason. There were no known
incompatibilities. We were releasing entirely new gems just to bump the
version stamp.

**Comment by reza-naq** [comment]
@dhh
>> I actually just removed the upper-bounds ...
Does this cover the deadlock on 'actionpack' too? 
Are when are you going to release these upper-bound changes? As you said this prevents many from upgrading to 5.1.0 for no reason. Maybe it should be given some sort of priority.

**Comment by dhh** [comment]
If you remove the upper-bound from simple_form, then your gemfile will
resolve, because Rails 5.1 will be allowed to require actionpack 5.1. Note,
this does not mean that simple_form necessarily works with Rails 5.1
without bugs! Simply that you'll be allowed to find out.

**Comment by reza-naq** [comment]
@ddh 
Got it. Thanks again

**Pull request merged**